scripts/ice-road-pipeline.py

```diff
@@ -124,15 +164,170 @@
+
+        # Load in the ASP translational-only matrix from pc-align tool
+        asp_matrix_fp = glob(f'{ice_dir}/results/pc-align-translation-only/temp-log*.txt')[0]
+        with open(asp_matrix_fp, 'r') as f:
+            lines = f.readlines()
+            for line in lines:
+                if "Translation vector (North-East-Down, meters): Vector3" in line: 
+                    n_e_d_shift = re.findall("[+-]?\d+\.\d+", line)
+                    n_e_d_shift = ','.join(n_e_d_shift)
+                    break
+
+        # get crs
+        ref_raster = rasterio.open(snow_tif)
+        crs = ref_raster.crs
+        pix_size, _ = ref_raster.res
+        crs = str(crs).split(":",1)[1]    
+        ras_meta = ref_raster.profile
+
+        # Compute slope and aspect from snow-on lidar
+        slope_fp = join(ice_dir, f'{basename(in_dir)}-snowon_slope.tif')
+        aspect_fp = join(ice_dir, f'{basename(in_dir)}-snowon_aspect.tif')
+        os.system(f'gdaldem slope -compute_edges {snow_tif} {slope_fp} -q')
+        os.system(f'gdaldem aspect -compute_edges -zero_for_flat {snow_tif} {aspect_fp} -q')
+        slope = rio.open_rasterio(slope_fp, masked=True)
+        aspect = rio.open_rasterio(aspect_fp, masked=True)
+
+        # Compute surface normal vector from snow-on lidar
+        ni_fp = join(ice_dir, f'{basename(in_dir)}-snowon_n_i.tif')
+        nj_fp = join(ice_dir, f'{basename(in_dir)}-snowon_n_j.tif')
+        nk_fp = join(ice_dir, f'{basename(in_dir)}-snowon_n_k.tif')
+        n_i = np.sin(np.radians(aspect)) * np.sin(np.radians(slope))
+        n_j = np.cos(np.radians(aspect)) * np.sin(np.radians(slope)) 
+        n_k = np.cos(np.radians(slope))
+        n_i.rio.to_raster(ni_fp)
+        n_j.rio.to_raster(nj_fp)
+        n_k.rio.to_raster(nk_fp)        
+        
+        # Calcs calibration stats from target
+        output_csv = f'{ssa_dir}/all-calibration-rfl.csv'
+        subprocess.call(["Rscript", 
+                          f"{scripts_dir}/las_ssa_cal.r", 
+                          cal_las, crs, shp_fp_rfl, n_e_d_shift, 
+                          output_csv, imu_data, str(pix_size)])
+        
+        # Read in cal data and estimate factor
+        df = pd.read_csv(output_csv)
+        median_rfl = df['rfl'].median()
+        road_cal_factor = known_rfl / median_rfl
+        road_cal_factor = str(road_cal_factor) #convert to str for R
+
+        # For each file in <in-dir> 
+        for f in os.listdir(in_dir):
+             
+            # Set names
+            base_las = os.path.basename(f)
+            las_name = os.path.splitext(base_las)[0]
+
+            # R is making a .DS_Store in in_dir, and I can't figure out how or where...
+            # So this will be just a quick check to continue to next iter if encountered.
+            if las_name == ".DS_Store" or las_name == "ice-road":
+                continue 
+
+            rfl_fp = f'{ssa_dir}/{las_name}-rfl.las'
+
+            # Getting a translated "LAS" file
```

Comment:
I don't totally understand what is happening in this step. Can you spell it out more in this comment?

Reply:
So there is a lot happening here, or rather a lot happening in `las_ssa_prep.R`. 

In summary, 

- 1 of the n LAS fight data is input
- LAS is shifted in X, Y, and Z direction based on ASP __pc_align__ with translation only
  - I note in the laz_align that using rotational cannot be done
- Heli IMU data (X,Y,Z,t) is used to build vector between surface and sensor --> used to estimate incidence angles.
- Reflectance is normalized + cleaned. 
- Finally, a new .LAS file is created where X and Y are position, but Z is the reflectance.

These n number of reflectance LAS are all compiled to create the mosaic raster using IDW.